### ISSUE TITLE
Value of source.workload.ui changed between 1.0 and 1.1

### ISSUE THREAD

**Issue by robertpanzer** [istio/istio#10873]
**Describe the bug**
I have a mixer rule using the listchecker adapter.
The listentry looks like this:
```yaml
apiVersion: config.istio.io/v1alpha2
kind: listentry
metadata:
  name: foo-workloadid
  namespace: istio-system
spec:
  value: source.workload.uid
```

The rule should ensure that only one certain deployment is allowed to use this route:
```yaml
apiVersion: config.istio.io/v1alpha2
kind: rule
metadata:
  name: subscribed-consumers-rule
  namespace: istio-system
spec:
  actions:
  - handler: foo.listchecker
    instances:
    - sr-workloadid.listentry
  match: request.host == "my-external.host" || request.host.startsWith("my-external.host:")
```

The listchecker looks like this:
```yaml
apiVersion: config.istio.io/v1alpha2
kind: listchecker
metadata:
  name: foo
  namespace: istio-system
spec:
  cachingInterval: 2m0s
  overrides:
  - istio://testns/workloads/testcurl
```

My routes are configured to route from a pod in the Istio cluster to the egress gateway.
The egress gateway routes the request to an external endpoint that is configured with a ServiceEntry.

If I run a curl request from a pod belonging to the deployment `testcurl` in the namespace `testns` with Istio 1.0.4 things are fine and the request passes.

If I run the same curl request from a pod with a recent Istio 1.1 snapshot the request is rejected with this error message:
```
PERMISSION_DENIED:foo.listchecker.istio-system:istio://istio-system/workloads/istio-egressgateway is not whitelisted
``` 

I added some Printfs to the mixer at `mixer/adapter/kubernetesenv/kubernetesenv.go` and for 1.0.4 I see this:
```
Fill source attributes for pod testcurl-f7dfd957-5glkz
Found testcurl istio://testns/workloads/testcurl
```

For 1.1 I see this:
```
Fill Source Attrs for Pod istio-egressgateway-5cfc47496b-dxmt6
Found istio-egressgateway
```

**Expected behavior**
I would expect that the attribute `source.workload.uid` contains the id of the deployment where the request originated, not the one where the mixer check is executed.

**Steps to reproduce the bug**
Configure a VirtualService that routes a request from the mesh gateway to the egress gateway and from there somewhere else, e.g. google.com.
Register a Deployment that gets the istio-proxy sidecar injected and that is able to send curl requests.
Create the rule as described.
Then try to send a request to the external provider.

**Version**
Istio built from the current release-1.1 branch, commit ee30c791a99acef108be825e70010b5f47da5254. 

**Installation**
```
helm template ~/go/src/istio.io/istio/install/kubernetes/helm/istio --name istio --namespace istio-system -f values-istio-auth.yaml  --set global.mtls.enabled=true --set global.tracing.enabled=true --set pilot.traceSampling=100.0 --set global.hub=istio --set global.tag=ee30c791a99acef108be825e70010b5f47da5254 --set tracing.enabled=true|k apply -f -
```

**Environment**
I am testing on Kubernetes 1.10.3 provided by Docker for Mac.

**Comment by kyessenov** [comment]
Gateways generally act as client-side proxies in istio. It gets confusing with the egress gateways since the source is ambiguous (it is either the gateway itself or whoever originated the request inside the mesh). I'm also not sure how hosts are carried over to the egress gateway. Can you please retrieve Envoy configurations for the egress and the test app using `istioctl proxy-config`? We can then understand the underlying Envoy's APIs.

CC'ing @rshriram @costinm who may know about egress differences between 1.0 and 1.1 off the top of one's head.

**Comment by robertpanzer** [comment]
Well, for requests to external destinations the egress gateway is the only place where I could apply a policy unless I want to enable client side mixer calls, which I definitely don't want as it seems to be a cluster wide setting.

When doing a call to a provider that is in the same cluster I can see that mixer identifies the correct consumer workload.
Only the mixer calls from the egress gateway seem to use the wrong workload id.

**Comment by robertpanzer** [comment]
Maybe it helps more trying to explain what I am trying to achieve:

I have routes to external systems configured that pass via the egress gateway.
At the same time I only want to allow usage of these routes only to certain workloads.

ServiceDependencies are a way that can achieve sth similar, but only as long as every workload has its own namespace, which cannot be assumed in general.

**Comment by kyessenov** [comment]
Could you please provide the networking config that you used? There may be differences depending on whether or not mTLS is used.

**Comment by mandarjog** [comment]
@kyessenov can you check the "inbound" |  "outbound" logic? I think there is a bug in that code inside proxy. It is detecting the egress proxy as "outbound" and then setting itself as source.

**Comment by jasminejaksic** [comment]
@kyessenov Are you working on this? Can I make you the assignee?

**Comment by robertpanzer** [comment]
Here is an example to reproduce the problem:

I have a simple configuration that forwards calls to http://www.google.com via the egress gateway to https://www.google.com.
Communication between the consumer and the egress gateway uses mTLS:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: egressgateway-google
  namespace: test
spec:
  selector:
    app: istio-egressgateway
    istio: egressgateway
  servers:
  - port:
      number: 80
      name: https-google
      protocol: HTTPS
    hosts:
    - www.google.com
    tls:
      mode: MUTUAL
      serverCertificate: /etc/certs/cert-chain.pem
      privateKey: /etc/certs/key.pem
      caCertificates: /etc/certs/root-cert.pem

---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: google-togateway
  namespace: test
spec:
  host: istio-egressgateway.istio-system.svc.cluster.local
  subsets:
  - name: togateway
    trafficPolicy:
      loadBalancer:
        simple: ROUND_ROBIN
      portLevelSettings:
      - port:
          number: 80
        tls:
          mode: ISTIO_MUTUAL
          sni: www.google.com

---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: google-toexternal
  namespace: test
spec:
  host: www.google.com
  trafficPolicy:
    connectionPool:
      tcp:
        connectTimeout: 5.000s
    tls:
      mode: SIMPLE
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: external-google
  namespace: test
spec:
  endpoints:
  - address: www.google.com
    ports:
      https: 443
  hosts:
  - www.google.com
  ports:
  - name: https
    number: 80
    protocol: HTTP
  resolution: DNS
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: google
  namespace: test
spec:
  gateways:
  - egressgateway-google
  - mesh
  hosts:
  - www.google.com
  http:
  - match:
    - gateways:
      - mesh
      port: 80
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        port:
          number: 80
        subset: togateway
      weight: 100
  - match:
    - gateways:
      - egressgateway-google
      port: 80
    route:
    - destination:
        host: www.google.com
        port:
          number: 80
      weight: 100
```

Now add a test consumer from which I am sending the test requests:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: testcurl
  namespace: sr-test
spec:
  selector:
    matchLabels:
      app: testcurl
  replicas: 1
  template:
    metadata:
      labels:
        app: testcurl
    spec:
      containers:
      - name: testcurl
        image: tutum/curl
        command:
          - sleep
          - "100000"
```

When we send a request from one of the pods of this Deployment we should get the google page:
```
kubectl exec -ti -n sr-test testcurl-... -- curl www.google.com
```

Now let's create the rules to allow nobody but this Deployment to make these requests:
```yaml
apiVersion: config.istio.io/v1alpha2
kind: checknothing
metadata:
  name: denyrequest
  namespace: istio-system
spec:
---
apiVersion: config.istio.io/v1alpha2
kind: denier
metadata:
  name: denyrequest
  namespace: istio-system
spec:
  validUseCount: 1
---
apiVersion: config.istio.io/v1alpha2
kind: rule
metadata:
  name: testrule
  namespace: istio-system
spec:
  match: source.workload.uid != "istio://sr-test/workloads/testcurl"
  actions:
    - handler: denyrequest.denier
      instances:
        - denyrequest.checknothing
```
So for every workload *except* the Deployment `testcurl` in the namespace `sr-test` the request should be denied.
And this works on Istio 1.0.

On 1.1 every request is rejected.
Only if I change the rule match to `source.workload.uid == "istio://istio-system/workloads/istio-egressgateway"` I can get the requests working again, but then of course from every workload.

**Comment by kyessenov** [comment]
Thank you. We found the culprit change in the proxy responsible for the regression.

**Comment by munrodg** [comment]
@mandarjog can you pick this one up?

**Comment by liminw** [comment]
@robertpanzer If the goal of using denier rule here is to provide security, instead of using source.workload.uid, we recommend to use source.principal (i.e., source service account). The reason is that source.principal (i.e., source service account) is strongly authenticated through mTLS, but source.workload.uid is not. source.workload.uid is ok to use for network routing/segmentation purpose, but not recommended for security purpose. 

In order to use source source account, you need to deploy the workloads that you want to assign different privilege to (e.g., testcurl in your example) with a separate service account.  

cc @wenchenglu @yangminzhu @incfly

**Comment by duderino** [comment]
@kyessenov @mandarjog what's left to be done here?  merge https://github.com/istio/istio/issues/10676 ?  Anything else?

**Comment by duderino** [comment]
@kyessenov @mandarjog would really love an update here

**Comment by mandarjog** [comment]
This is done, it is available in 1.1 now.
